hbase-external-blockcache/src/test/java/org/apache/hadoop/hbase/io/hfile/TestMemcachedBlockCache.java

```diff
@@ -0,0 +1,124 @@
+import com.thimbleware.jmemcached.storage.hash.ConcurrentLinkedHashMap;
+import com.thimbleware.jmemcached.storage.hash.ConcurrentLinkedHashMap.EvictionPolicy;
+import java.net.InetSocketAddress;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.HBaseClassTestRule;
+import org.apache.hadoop.hbase.HBaseTestingUtil;
+import org.apache.hadoop.hbase.HConstants;
+import org.apache.hadoop.hbase.Waiter;
+import org.apache.hadoop.hbase.io.hfile.CacheTestUtils.HFileBlockPair;
+import org.apache.hadoop.hbase.testclassification.IOTests;
+import org.apache.hadoop.hbase.testclassification.LargeTests;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+@Category({ IOTests.class, LargeTests.class })
+public class TestMemcachedBlockCache {
+
+  @ClassRule
+  public static final HBaseClassTestRule CLASS_RULE =
+    HBaseClassTestRule.forClass(TestMemcachedBlockCache.class);
+
+  static MemCacheDaemon<? extends CacheElement> MEMCACHED;
+  static MemcachedBlockCache CACHE;
+
+  @Before
+  public void before() throws Exception {
+    MEMCACHED.getCache().flush_all();
+    assertEquals("Memcache is not empty", MEMCACHED.getCache().getCurrentItems(), 0);
+  }
+
+  @BeforeClass
+  public static void setup() throws Exception {
+    int port = HBaseTestingUtil.randomFreePort();
+    MEMCACHED = createDaemon(port);
+    Configuration conf = new Configuration();
+    conf.set("hbase.cache.memcached.servers", "localhost:" + port);
+    CACHE = new MemcachedBlockCache(conf);
+  }
+
+  @AfterClass
+  public static void tearDown() throws Exception {
+    if (MEMCACHED != null) {
+      MEMCACHED.stop();
+    }
+  }
+
+  @Test
+  public void testCache() throws Exception {
+    final int NUM_BLOCKS = 10;
+    HFileBlockPair[] blocks =
+      CacheTestUtils.generateHFileBlocks(HConstants.DEFAULT_BLOCKSIZE, NUM_BLOCKS);
+    for (int i = 0; i < NUM_BLOCKS; i++) {
+      CACHE.cacheBlock(blocks[i].getBlockName(), blocks[i].getBlock());
+    }
+    Waiter.waitFor(new Configuration(), 10000,
+      () -> MEMCACHED.getCache().getCurrentItems() == NUM_BLOCKS);
+  }
+
+  @Test
+  public void testEviction() throws Exception {
+    final int NUM_BLOCKS = 10;
+    HFileBlockPair[] blocks =
+      CacheTestUtils.generateHFileBlocks(HConstants.DEFAULT_BLOCKSIZE, NUM_BLOCKS);
+    for (int i = 0; i < NUM_BLOCKS; i++) {
+      CACHE.cacheBlock(blocks[i].getBlockName(), blocks[i].getBlock());
+    }
+    Waiter.waitFor(new Configuration(), 10000,
+      () -> MEMCACHED.getCache().getCurrentItems() == NUM_BLOCKS);
+    for (int i = 0; i < NUM_BLOCKS; i++) {
+      CACHE.evictBlock(blocks[i].getBlockName());
+    }
+    Waiter.waitFor(new Configuration(), 10000, () -> MEMCACHED.getCache().getCurrentItems() == 0);
+  }
+
+  private static MemCacheDaemon<? extends CacheElement> createDaemon(int port) {
```

Comment:
Looks like we didn't have tests so far that runs memcached locally?

Reply:
That is correct. This is the first unit test for this feature.